pkg/reconciler/pipelinerun/timeout.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	clientset "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	jsonpatch "gomodules.xyz/jsonpatch/v2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/apis"
+)
+
+var timeoutTaskRunPatchBytes []byte
+
+func init() {
+	var err error
+	timeoutTaskRunPatchBytes, err = json.Marshal([]jsonpatch.JsonPatchOperation{{
+		Operation: "add",
+		Path:      "/spec/status",
+		Value:     v1beta1.TaskRunSpecStatusTimedOut,
+	}})
+	if err != nil {
+		log.Fatalf("failed to marshal TaskRun timed out patch bytes: %v", err)
+	}
+}
+
+// timeoutPipelineRun marks the PipelineRun as timed out and any resolved TaskRun(s) too.
+func timeoutPipelineRun(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, clientSet clientset.Interface) error {
+	errs := timeoutPipelineTaskRuns(ctx, logger, pr, clientSet)
+
+	// If we successfully timed out all the TaskRuns and Runs, we can consider the PipelineRun timed out.
+	if len(errs) == 0 {
+		reason := v1beta1.PipelineRunReasonTimedOut.String()
+
+		pr.Status.SetCondition(&apis.Condition{
+			Type:    apis.ConditionSucceeded,
+			Status:  corev1.ConditionFalse,
+			Reason:  reason,
+			Message: fmt.Sprintf("PipelineRun %q failed to finish within %q", pr.Name, pr.PipelineTimeout(ctx).String()),
+		})
+		// update pr completed time
+		pr.Status.CompletionTime = &metav1.Time{Time: time.Now()}
+	} else {
+		e := strings.Join(errs, "\n")
+		// Indicate that we failed to time out the PipelineRun
+		pr.Status.SetCondition(&apis.Condition{
+			Type:    apis.ConditionSucceeded,
+			Status:  corev1.ConditionUnknown,
+			Reason:  ReasonCouldntTimeOut,
+			Message: fmt.Sprintf("PipelineRun %q was timed out but had errors trying to time out TaskRuns and/or Runs: %s", pr.Name, e),
+		})
+		return fmt.Errorf("error(s) from timing out TaskRun(s) from PipelineRun %s: %s", pr.Name, e)
+	}
+	return nil
+}
+
+// timeoutPipelineTaskRuns patches `TaskRun` and `Run` with timed out status
+func timeoutPipelineTaskRuns(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, clientSet clientset.Interface) []string {
```

Comment:
this function name is confusing - made me think we were timing out `taskruns` but it's also timing out `runs`

can we split it into two functions - `timeoutTaskRuns` and `timeoutRuns`? then we call both of them in `timeoutPipelineRun`?

Reply:
I'd actually go with renaming it to just `timeoutPipelineTasks` and keeping `TaskRun` and `Run` timeout in here - the name was just copy-pasted from `cancel.go` and renamed from `cancelPipelineTaskRuns`.